llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -3382,6 +3418,17 @@ Instruction *InstCombinerImpl::visitAllocSite(Instruction &MI) {
           eraseInstFromFunction(*I);
           Users[i] = nullptr; // Skip examining in the next loop.
         }
+        if (auto *MTI = dyn_cast<MemTransferInst>(I)) {
+          if (KnowInitZero && getUnderlyingObject(MTI->getRawDest()) != &MI) {
+            IRBuilderBase::InsertPointGuard Guard(Builder);
+            Builder.SetInsertPoint(MTI);
+            auto *M = Builder.CreateMemSet(MTI->getRawDest(),
+                    ConstantInt::get(Type::getInt8Ty(MI.getContext()), 0),
+                    MTI->getLength(),
+                    MTI->getDestAlign());
+            M->copyMetadata(*MTI, LLVMContext::MD_DIAssignID);
```

Comment:
Should probably copy all the metadata? E.g. to preserve AA metadata.

Reply:
I was wondering about that too. The current MemCpyOptimizer pass usually drops all metadata information (which is where I copied this from to be consistent) but I couldn't really justify that to myself either.